From 4647add0c6a2cd4ff2a281b1fd4cb7dafa196161 Mon Sep 17 00:00:00 2001
From: dev4u <dev4u@users.noreply.github.com>
Date: Thu, 17 Mar 2022 08:15:06 +0800
Subject: [PATCH 07/13] Update client.rs (#4)

Fix the issue that the connection creation fails when the network changes
---
 src/client.rs | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

--- a/src/client.rs
+++ b/src/client.rs
@@ -115,7 +115,26 @@ async fn transfer(
     let new_conn = endpoint
         .connect(*remote, &host)?
         .await
-        .map_err(|e| anyhow!("failed to connect: {}", e))?;
+        .map_err(|e| {
+            if e == ConnectionError::TimedOut {
+                let socket = if cfg!(target_os = "windows") {
+                    std::net::UdpSocket::bind("0.0.0.0:0").unwrap()
+                } else {
+                    std::net::UdpSocket::bind("[::]:0").unwrap()
+                };
+                let addr = socket.local_addr().unwrap();
+                let ret = endpoint.rebind(socket);
+                match ret {
+                    Ok(_) => {
+                        info!("rebinding to: {}", addr);
+                    }
+                    Err(e) => {
+                        error!("rebind fail: {:?}", e);
+                    }
+                }
+            }
+            anyhow!("failed to connect: {:?}", e)
+        }).unwrap();
 
     let quinn::NewConnection {
         connection: conn, ..
